community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/TransactionStateMachine.java

```diff
@@ -95,6 +95,14 @@ public void streamResult( ThrowingConsumer<BoltResult, Exception> resultConsumer
         }
     }
 
+    /**
+     * Rollback and close transaction. Move back to {@link State#AUTO_COMMIT}.
+     * <p>
+     * <b>Warning:</b>This method should only be called by the bolt worker thread during it's regular message
+     * processing. It is wrong to call it from a different thread because kernel transactions are not thread-safe.
```

Comment:
We can propagate bolt worker thread id from `RunnableBoltWorker` to `BoltStateMachine` and from there to this `TransactionStateMachine`. Then we can probably add runtime checks like this in all methods:

```java
if(Thread.currentThread() != boltWorkerThread)
    throw new IllegalStateException();
```

Do you think it is worth doing? Maybe as a separate PR?